### PR TITLE
Create and Update Events for DRPC

### PR DESCRIPTION
**Pull request by jmishrarh**
Added few DRPC events in the code for 
1. Deployment 2. Failover and 3. Relocate
Yet add events for metro/sync  failure /relocate scenarios.
Log snippet 
2428:2022-11-23T16:16:16.083+0530       DEBUG   events  Preparing for exectuion {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "DRPCInitiating"}
2429:2022-11-23T16:16:16.083+0530       DEBUG   events  Deploying the application and VRG       {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "DRPCDeploying"}
2430:2022-11-23T16:16:16.083+0530       DEBUG   events  Creating VRG ManifestWork       {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "CreateVRGManifestWork"}
2435:2022-11-23T16:16:16.085+0530       DEBUG   events  Updating userPlacementRule user-placement-rule  {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "UpdateUserPlacementRule"}
2439:2022-11-23T16:16:16.087+0530       DEBUG   events  Successfully deployed the application and VRG   {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "DRPCDeploySuccess"}
2465:2022-11-23T16:16:16.094+0530       DEBUG   events  Initial Deployment completed    {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"548"}, "reason": "DRPCDeploySuccess"}
2493:2022-11-23T16:16:16.099+0530       DEBUG   events  Successfully deployed the application and VRG   {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"554"}, "reason": "DRPCDeploySuccess"}
2494:2022-11-23T16:16:16.099+0530       DEBUG   events  Initial Deployment completed    {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"554"}, "reason": "DRPCDeploySuccess"}
2519:2022-11-23T16:16:16.105+0530       DEBUG   events  Successfully deployed the application and VRG   {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"554"}, "reason": "DRPCDeploySuccess"}
2520:2022-11-23T16:16:16.105+0530       DEBUG   events  Initial Deployment completed    {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"554"}, "reason": "DRPCDeploySuccess"}
2563:2022-11-23T16:16:16.118+0530       DEBUG   events  Preparing for exectuion {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"560"}, "reason": "DRPCInitiating"}
2566:2022-11-23T16:16:16.118+0530       DEBUG   events  Failing over the application and VRG    {"type": "Warning", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"560"}, "reason": "DRPCFailingOver"}
2567:2022-11-23T16:16:16.118+0530       DEBUG   events  Creating VRG ManifestWork       {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"560"}, "reason": "CreateVRGManifestWork"}
2571:2022-11-23T16:16:16.120+0530       DEBUG   events  Waiting for PV restore to complete...)  {"type": "Warning", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"560"}, "reason": "DRPCClusterSwitchFailed"}
2941:2022-11-23T16:16:26.131+0530       DEBUG   events  Updating userPlacementRule user-placement-rule  {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"571"}, "reason": "UpdateUserPlacementRule"}
2947:2022-11-23T16:16:26.133+0530       DEBUG   events  Successfully failedover the application and VRG {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"571"}, "reason": "DRPCFailoverSuccess"}
2977:2022-11-23T16:16:26.140+0530       DEBUG   events  Cleanup starting on Secondary   {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"576"}, "reason": "StartingSecondaryCleanup"}
2981:2022-11-23T16:16:26.142+0530       DEBUG   events  Failed to cleanup on Secondary  {"type": "Warning", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"576"}, "reason": "CleanupSecondaryUnsuccessful"}
3010:2022-11-23T16:16:26.148+0530       DEBUG   events  Cleanup starting on Secondary   {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"580"}, "reason": "StartingSecondaryCleanup"}
3023:2022-11-23T16:16:26.153+0530       DEBUG   events  Cleanup completed       {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"580"}, "reason": "CleanupSecondarySuccessful"}
3055:2022-11-23T16:16:26.159+0530       DEBUG   events  Cleanup starting on Secondary   {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"585"}, "reason": "StartingSecondaryCleanup"}
3086:2022-11-23T16:16:26.170+0530       DEBUG   events  Updating userPlacementRule user-placement-rule  {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"588"}, "reason": "UpdateUserPlacementRule"}
3096:2022-11-23T16:16:26.172+0530       DEBUG   events  Successfully failedover the application and VRG {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"588"}, "reason": "DRPCFailoverSuccess"}

Signed-off-by: Jolly Mishra <jmishra@redhat.com>

### CODE REVIEW

File: controllers/drplacementcontrol.go [hunk 1]
```diff
@@ -103,6 +103,8 @@ func (d *DRPCInstance) RunInitialDeployment() (bool, error) {
 
 	const done = true
 
+	d.reportEvent(d.getLastDRState())
```
Comment (BenamarMk) [inline controllers/drplacementcontrol.go:106]: Why is this here? `setDRState` will always call it once the state changes.

**Review by BenamarMk — APPROVED** [review]
Based on the log snippet, it looks good to me.

File: controllers/drplacementcontrol.go (hunk 1 above)
Comment (jmishrarh) [reply controllers/drplacementcontrol.go:106]: Yes. Correct. I noticed that there are many unknown events reported by default for example
2022-11-23T16:16:16.083+0530    DEBUG   events  next state not known    {"type": "Warning", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "unknown state"}
I tried to add more info with event DRPCInitiating instead of unknown state which kind of confusing.  That's the reason I called the reportEvent. we still have some intermediate state where 'unknown state' reported. 
If this is making more verbose I can remove and let it be as it was with couple of unknown state events.
2022-11-23T16:16:16.083+0530    DEBUG   events  Preparing for exectuion {"type": "Normal", "object": {"kind":"DRPlacementControl","namespace":"app-namespace","name":"app-volume-replication-test","uid":"a79c9853-b867-46e2-a11f-e7b7b3ab2859","apiVersion":"ramendr.openshift.io/v1alpha1","resourceVersion":"542"}, "reason": "DRPCInitiating"}

File: controllers/util/events.go [hunk 2]
```diff
@@ -71,24 +75,60 @@ const (
...
+	EventReasonVRGManifestWorkCreationFailed = "CreateVRGManifestWorkFailed"
+
+	// EventReasonUpdateUserPlacementRule is generated to update the user PlacementRule
+	EventReasonUpdateUserPlacementRule = "UpdateUserPlacementRule"
+
+	// EventReasonUpdateUserPlacementRuleFailed is generated when
+	// DRPC fails to update the user PlacementRule
+	EventReasonUpdateUserPlacementRuleFailed = "UpdateUserPlacementRuleFailed"
+
+	// EventReasonDeployFail is generic, can we re-use it with appropriate failure reason?
```
Comment (jmishrarh) [inline controllers/util/events.go:92]: EventReasonDeployFail is generic, can we re-use it with appropriate failure message  or use more specific reason like "EventReasonVRGManifestWorkCreationFailed" instead?

File: controllers/drplacementcontrol.go [hunk 3]
```diff
@@ -162,6 +164,9 @@ func (d *DRPCInstance) RunInitialDeployment() (bool, error) {
 
 	d.setProgression(rmn.ProgressionCompleted)
 
+	rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeNormal,
```
Comment (ShyamsundarR) [inline controllers/drplacementcontrol.go:167]: This event is reported by setDRState within `startDeploying` or if empty just above in line 157. This would be a repeat event and also not required.

File: controllers/drplacementcontrol.go [hunk 4]
```diff
@@ -264,6 +269,7 @@ func (d *DRPCInstance) startDeploying(homeCluster, homeClusterNamespace string)
 
 	err = d.updateUserPlacementRule(homeCluster, homeClusterNamespace)
 	if err != nil {
+		err := fmt.Errorf("failed to update UserPlacementRule %s", d.userPlacementRule.Name)
```
Comment (ShyamsundarR) [inline controllers/drplacementcontrol.go:272]: Why do we need this? `updateUserPlacementRuleStatus` returns the said error string and also the error from the API server that was received.

File: controllers/drplacementcontrol.go [hunk 5]
```diff
@@ -307,6 +313,9 @@ func (d *DRPCInstance) RunFailover() (bool, error) {
 		d.setDRPCCondition(&d.instance.Status.Conditions, rmn.ConditionAvailable, d.instance.Generation,
 			d.getConditionStatusForTypeAvailable(), string(d.instance.Status.Phase), msg)
 
+		rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeWarning,
```
Comment (ShyamsundarR) [inline controllers/drplacementcontrol.go:316]: We potentially need an event if `isPlacementNeedsFixing` fails, or even if it executed a fixup, so that we know a state was updated/fixed.

File: controllers/drplacementcontrol.go [hunk 6]
```diff
@@ -487,6 +496,9 @@ func (d *DRPCInstance) RunRelocate() (bool, error) {
 		d.setDRPCCondition(&d.instance.Status.Conditions, rmn.ConditionAvailable, d.instance.Generation,
 			d.getConditionStatusForTypeAvailable(), string(d.instance.Status.Phase), err.Error())
 
+		rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeWarning,
```
Comment (ShyamsundarR) [inline controllers/drplacementcontrol.go:499]: Event message can be return of `validateAndSelectCurrentPrimary` as the messages are granular from the function, the current message is generic and will not give details to the user.

File: controllers/drplacementcontrol.go (hunk 4 above)
Comment (ShyamsundarR) [reply controllers/drplacementcontrol.go:272]: We also seem to have added an event in `updateUserPlacementRuleStatus` which would hence be a duplicate.

**Review by ShyamsundarR — COMMENTED** [review]
Started an initial review, some comments to consider. We need to be aware of setDRState as @BenamarMk points out and avoid duplicate events.

File: controllers/drplacementcontrol.go (hunk 5 above)
Comment (jmishrarh) [reply controllers/drplacementcontrol.go:312]: Done

**Comment by ShyamsundarR** [comment]
Closing as stale, part of general cleanup. Please reopen and update if still required.
